build-tools/packages/build-cli/src/library/packageExports.ts

```diff
@@ -0,0 +1,140 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+
+import type { Logger, PackageJson } from "@fluidframework/build-tools";
+
+import { ApiTag } from "./apiTag";
+
+export interface ExportData {
+	/**
+	 * Location of file relative to package
+	 */
+	relPath: string;
+	/**
+	 * Conditions required for export access; in hierarchy order
+	 */
+	conditions: string[];
+	/**
+	 * Export is only .d.ts file
+	 */
+	isTypeOnly: boolean;
+}
+export type Node10CompatExportData = Omit<ExportData, "conditions">;
+
+/**
+ * Only the value types of exports that are records.
+ */
+type ExportsRecordValue = Exclude<Extract<PackageJson["exports"], object>, unknown[]>;
+
+function findTypesPathMatching(
+	mapQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined>,
+	exports: ExportsRecordValue,
+	conditions: string[],
+): (ExportData & { apiTagLevel: ApiTag | undefined }) | undefined {
+	for (const [entry, value] of Object.entries(exports)) {
+		if (typeof value === "string") {
+			if (entry === "types") {
+				for (const [key, apiTagLevel] of mapQueryPathToApiTagLevel.entries()) {
+					// eslint-disable-next-line max-depth
+					if (
+						typeof key === "string"
+							? path.resolve(value) === path.resolve(key)
+							: key.test(value)
+					) {
+						const isTypeOnly = !(
+							"default" in exports ||
+							"import" in exports ||
+							"require" in exports
+						);
+						return { apiTagLevel, relPath: value, conditions, isTypeOnly };
+					}
+				}
+			}
+		} else if (value !== null) {
+			if (Array.isArray(value)) {
+				continue;
+			}
+			const deepFind = findTypesPathMatching(mapQueryPathToApiTagLevel, value, [
+				...conditions,
+				entry,
+			]);
+			if (deepFind !== undefined) {
+				return deepFind;
+			}
+		}
+	}
+
+	return undefined;
+}
+
+export function queryOutputMapsFromPackageExports(
+	packageJson: PackageJson,
+	mapQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined>,
+	node10TypeCompat: boolean,
+	logger?: Logger,
+): {
+	mapApiTagLevelToOutput: Map<ApiTag, ExportData>;
+	mapNode10CompatExportPathToData: Map<string, Node10CompatExportData>;
+} {
+	const mapApiTagLevelToOutput = new Map<ApiTag, ExportData>();
+	const mapNode10CompatExportPathToData = new Map<string, Node10CompatExportData>();
+
+	const { exports } = packageJson;
+	if (typeof exports !== "object" || exports === null) {
+		throw new Error('no valid "exports" within package properties');
+	}
+
+	if (Array.isArray(exports)) {
+		// eslint-disable-next-line unicorn/prefer-type-error
+		throw new Error(`required entrypoints cannot be generated for "exports" array`);
+	}
+
+	// Iterate through exports looking for properties with values matching keys in map.
```

Comment:
Seems very similar in intent to https://github.com/microsoft/FluidFramework/blob/f44e1f0b950efc624c005545f1ea769d11860535/build-tools/packages/build-cli/src/commands/generate/typetests.ts#L181-L186

Reply:
I didn't know about `resolve.exports`. I think I might be gathering more information during the processing than it would provide, but I can check it out. I don't think I want to block on getting to that.

Reply:
Actually, `resolve.exports` doesn't appear helpful. The command this supports works backwards from a check of file path.